Use strongly typed Ids

psincf opened the pull request:
fixes #61

kvark commented on wgpu-native/src/command/compute.rs:
```diff
@@ -33,11 +33,11 @@ impl<B: hal::Backend> ComputePass<B> {
 
 #[no_mangle]
 pub extern "C" fn wgpu_compute_pass_end_pass(pass_id: ComputePassId) -> CommandBufferId {
-    let pass = HUB.compute_passes.unregister(pass_id);
```
would it be feasible to have `compute_passes` (and other things inside `HUB`) to be strongly typed by the ID?

psincf replied:
That's what I tried to do first, but I have some hard time with it.
It would certainly be better.

kvark commented on wgpu-native/src/lib.rs:
```diff
@@ -168,46 +168,100 @@ pub struct ByteArray {
+}
+
+use hub::{Index, Epoch, NewId, Id};
+macro_rules! new_id {
```
why can't we have a single macro instead of 2?

psincf replied:
I agree, I was just adding a new macro since I needed a new one, but we can merge both into one

kvark commented on wgpu-native/src/lib.rs:
```diff
@@ -168,46 +168,100 @@ pub struct ByteArray {
+transparent!(pub struct SurfaceId(hub::Id););
 type SurfaceHandle = Surface<back::Backend>;
-pub type SwapChainId = SurfaceId;
+transparent!(pub struct SwapChainId(SurfaceId););
```
I think it should be `hub::Id` here, although it depends on whether we'll manage to have the strong typing integrated into `HUB` itself.

kvark commented on wgpu-native/src/track.rs:
```diff
@@ -274,37 +274,37 @@ impl<I: NewId, U: Copy + GenericUsage + BitOr<Output = U> + PartialEq> Tracker<I
 }
 
-impl<U: Copy + GenericUsage + BitOr<Output = U> + PartialEq> Tracker<Id, U> {
+impl<I: ToId + NewId, U: Copy + GenericUsage + BitOr<Output = U> + PartialEq> Tracker<I, U> {
```
Can we have everything in `IdLike` trait?

kvark commented on wgpu-native/src/lib.rs:
```diff
@@ -168,46 +168,100 @@ pub struct ByteArray {
+    )
+}
+
+pub trait ToId {
```
shouldn't be public

kvark commented on wgpu-native/src/lib.rs:
```diff
@@ -168,46 +168,100 @@ pub struct ByteArray {
+}
+
+pub trait ToId {
+    fn id(&self) -> Id;
```
let's call it `raw_id()`

psincf commented:
@kvark 
Finally found a way to integrate Strong typing directly into HUB :), It should be better than relying on hub::Id each time. 
Need be cleaned a bit (merged macros, function name, trait name, and merged trait with the `IdLike` trait you proposed ).
Also `SwapChainId`, `CommandEncoderId`, and `QueueId` are just aliased types to their equivalent.

kvark commented:
Thank you! currently, formatting changes get in the way of reviewing this. Please rebase so that we can continue.

psincf commented:
I think rebase have been done correctly.. I'm still discovering git so don't hesitate to check!
Need to be cleaned a bit as said [here](https://github.com/gfx-rs/wgpu/pull/136#issuecomment-487192505).

kvark commented on wgpu-native/src/command/mod.rs:
```diff
@@ -99,8 +99,8 @@ impl CommandBufferHandle {
         stitch: Stitch,
-        buffer_guard: &Storage<BufferHandle>,
-        texture_guard: &Storage<TextureHandle>,
+        buffer_guard: &Storage<BufferHandle, crate::BufferId>,
```
let's be consistent with use of other *Id types and just `use` it at the top instead of having absolute paths in the code

kvark commented on wgpu-native/src/device.rs:
```diff
@@ -802,7 +802,7 @@ pub fn texture_create_view(
 }
 
-pub fn device_track_view(texture_id: TextureId, view_id: BufferId, ref_count: RefCount) {
+pub fn device_track_view(texture_id: TextureId, view_id: TextureViewId, ref_count: RefCount) {
```
ooops:)

kvark commented on wgpu-native/src/hub.rs:
```diff
@@ -87,107 +87,112 @@ impl IdentityManager {
     identity: Mutex<IdentityManager>,
-    data: RwLock<Storage<T>>,
+    data: RwLock<Storage<T, I>>,
+    _phantom: std::marker::PhantomData<&'static I>,
```
I don't think this phantom member is needed here

kvark commented on wgpu-native/src/lib.rs:
```diff
@@ -168,46 +168,157 @@ pub struct ByteArray {
+    )
+}
+
+pub trait ToId {
```
can we call it `TypedId`?

kvark commented on wgpu-native/src/lib.rs:
```diff
@@ -168,46 +168,157 @@ pub struct ByteArray {
+}
+
+pub trait ToId {
+    fn id(&self) -> hub::Id;
```
can we call it `raw()`?

kvark commented on wgpu-native/src/lib.rs:
```diff
@@ -168,46 +168,157 @@ pub struct ByteArray {
+    )
+}
+
+macro_rules! from_id {
```
can we merge these two macros into one?

kvark commented on wgpu-native/src/lib.rs:
```diff
@@ -168,46 +168,157 @@ pub struct ByteArray {
+                $i(id)
+            }
+
+            fn index(&self) -> Index {
```
I think we can omit it and just make `trait NewId: TypedId`

psincf replied:
I am not sure to understand, you propose to integrate the `NewId` into the `TypedId` trait?

kvark replied:
Nevermind, I think your approach is fine

kvark commented on wgpu-native/src/lib.rs:
```diff
@@ -168,46 +168,157 @@ pub struct ByteArray {
 // Swap chain
-pub type SurfaceId = hub::Id;
+transparent!(pub struct SurfaceId(hub::Id););
+to_id!(SurfaceId);
```
everything should be done by a single macro, i.e. `define_id(SurfaceId)` can both declare the type and implement the needed things for it

kvark commented:
This looks to be ready now. @grovesNL do you want to have a look as well?

grovesNL commented on wgpu-native/src/hub.rs:
```diff
@@ -87,107 +105,109 @@ impl IdentityManager {
 }
 
-pub struct Storage<T> {
+pub struct Storage<T, I:'static + TypedId> {
```
Is the `'static` bound needed for these generics? It seems like we could just have `PhantomData<I>`

psincf replied:
Yes you are right, I changed it.

psincf commented:
> I also think there might be a couple small changes needed to pass CI – it looks like I get some build errors when enabling `serde` using `--features remote`. It seems like Travis CI doesn't run on draft pull requests?

Apparently, the same error is raising if we run `cargo check` on the main folder.
But it works fine if we run it in examples.

kvark commented on wgpu-native/src/lib.rs:
```diff
@@ -166,46 +169,117 @@ pub struct ByteArray {
+macro_rules! typed_id {
+    ($i:ident) => ( 
+        impl From<Id> for $i {
+            fn from(id:Id) -> $i {
```
let's have this method just as `fram_raw` instead of requiring `From<Id>`

psincf replied:
Now that `IdentityManager` is strongly typed, a `from_raw()` method is not necessary so I delete it

psincf commented:
> 
> Yes, that would be good.

When making `IdentityManager` strongly typed, it became unnecessary to be able to use `Id` directly except for the `TypedId` struct, so I moved it from _hub.rs_ to _lib.rs_, and made it private. Also the `raw()` method is now only callable in _lib.rs_
`Index` and `Epoch` are moved in _lib.rs_.
I deleted `NewId` trait, which is making the code cleaner in my opinion, and less redundant, because in the previous commits, `NewId` was both implemented for `Id` and all the Typed Ids. Now there is only `TypedId` trait.
